Generate scenario from trace

achamayou opened the pull request:
Very basic script, using functionality introduced by https://github.com/tlaplus/tlaplus/issues/853, try and translate a trace to a raft scenario, for the purpose of executing counter-examples on the implementation.

Example run, using the following spec break:
```
LogMatchingInv ==
    \A i, j \in Servers : i /= j => Len(log[i]) < 10
```

```
./tlc.sh -workers auto -config consensus/MCccfraftWithReconfig.cfg consensus/MCccfraft.tla -dumpTrace tla MCccfraftWithReconfig.trace.tla -dumpTrace json MCccfraftWithReconfig.json

$ python3 trace2scen.py MCccfraftWithReconfig.json > scen
start_node,n1
emit_signature,2
# MCClientRequest MCccfraft:65
replicate,2,42
assert_is_primary,n1
# MCSignCommittableMessages MCccfraft:73
emit_signature,2
assert_is_primary,n1
# ChangeConfiguration ccfraft:660
replicate_new_configuration,2,n1,n2
assert_is_primary,n1
# MCSignCommittableMessages MCccfraft:73
emit_signature,2
assert_is_primary,n1
# MCClientRequest MCccfraft:65
replicate,2,42
assert_is_primary,n1
# MCSignCommittableMessages MCccfraft:73
emit_signature,2
assert_is_primary,n1
# ChangeConfiguration ccfraft:660
replicate_new_configuration,2,n2
assert_is_primary,n1
# MCSignCommittableMessages MCccfraft:73
emit_signature,2
assert_is_primary,n1
assert_commit_idx,n1,2
assert_commit_idx,n2,0
```

It's not straightforward to do this correctly, because the spec is set up differently from the driver: individual message delivery and consumption is contained in steps, whereas the driver has a separate dispatch action. This difference in granularity means that we probably can't reproduce every scenario precisely, but we can get quite close most of the time still.

lemmy commented:
Some discoveries:

### Node with TLA+ state "None" never created in scenario causing `core dump` when, e.g., asserting its state

```shell
root@codespaces-a47fa6:/workspaces/CCF/tests/raft_scenarios# ../../build/raft_driver MCccfraftAtomicReconfig 
2023-12-19T21:59:35.734666Z        100 [info ][raft] ./src/consensus/aft/raft.h:491 | n[n1] | None | Active | Election timer has become active
2023-12-19T21:59:35.734735Z        100 [debug][raft] /src/consensus/aft/raft.h:1820 | n[n1] | None | Active | Election index is 0 in term 2
<RaftDriver>  n[n1]->>n[n1]: [KV] initialising in term 2
2023-12-19T21:59:35.734754Z        100 [info ][raft] /src/consensus/aft/raft.h:1844 | n[n1] | Leader | Active | Becoming leader n[n1]: 2
2023-12-19T21:59:35.734761Z        100 [debug][raft] /src/consensus/aft/raft.h:2185 | n[n1] | Leader | Active | Starting commit
<RaftDriver>  n[n1]->>n[n1]: replicate 2.1 = [0 bytes]  [reconfiguration]
2023-12-19T21:59:35.735182Z        100 [debug][raft] ./src/consensus/aft/raft.h:567 | n[n1] | Leader | Active | Replicating 1 entries
2023-12-19T21:59:35.735193Z        100 [debug][raft] ./src/consensus/aft/raft.h:591 | n[n1] | Leader | Active | Replicated on leader n[n1]: 1 (1 hooks)
2023-12-19T21:59:35.735199Z        100 [debug][raft] ./src/consensus/aft/raft.h:451 | n[n1] | Leader | Active | Configurations: add new configuration at 1: {{n[n1]}}
<RaftDriver>  n[n1]->>n[n1]: [ledger] appending: 2.1=[68 bytes] {"data":"eyJuMS
2023-12-19T21:59:35.735209Z        100 [debug] ../src/consensus/aft/impl/state.h:39 | Updating view to: 2 at version: 1
2023-12-19T21:59:35.735214Z        100 [debug] ../src/consensus/aft/impl/state.h:56 | Resulting views: 1, 1
<RaftDriver>  Note over n1: Node n1 created
<RaftDriver>  n[n1]->>n[n1]: replicate 2.2 = [9 bytes] signature [raw]
2023-12-19T21:59:35.735237Z        100 [debug][raft] ./src/consensus/aft/raft.h:567 | n[n1] | Leader | Active | Replicating 1 entries
2023-12-19T21:59:35.735244Z        100 [debug][raft] ./src/consensus/aft/raft.h:591 | n[n1] | Leader | Active | Replicated on leader n[n1]: 2 committable (0 hooks)
2023-12-19T21:59:35.735249Z        100 [debug][raft] ./src/consensus/aft/raft.h:613 | n[n1] | Leader | Active | membership: Active leadership: Leader
<RaftDriver>  n[n1]->>n[n1]: [ledger] appending: 2.2=[36 bytes] {"data":"c2lnbm
2023-12-19T21:59:35.735260Z        100 [debug] ../src/consensus/aft/impl/state.h:39 | Updating view to: 2 at version: 2
2023-12-19T21:59:35.735264Z        100 [debug] ../src/consensus/aft/impl/state.h:56 | Resulting views: 1, 1
2023-12-19T21:59:35.735269Z        100 [debug][raft] /src/consensus/aft/raft.h:2128 | n[n1] | Leader | Active | In update_commit, new_commit_idx: 2, last_idx: 2
2023-12-19T21:59:35.735274Z        100 [debug][raft] /src/consensus/aft/raft.h:2185 | n[n1] | Leader | Active | Starting commit
2023-12-19T21:59:35.735279Z        100 [debug][raft] /src/consensus/aft/raft.h:2203 | n[n1] | Leader | Active | Compacting...
<RaftDriver>  n[n1]->>n[n1]: [KV] compacting to 2
2023-12-19T21:59:35.735288Z        100 [debug][raft] /src/consensus/aft/raft.h:2207 | n[n1] | Leader | Active | Commit on n[n1]: 2
<RaftDriver>  n[n1]->>n[n1]: replicate 2.3 = [2 bytes] 42 [raw]
2023-12-19T21:59:35.735303Z        100 [debug][raft] ./src/consensus/aft/raft.h:567 | n[n1] | Leader | Active | Replicating 1 entries
2023-12-19T21:59:35.735308Z        100 [debug][raft] ./src/consensus/aft/raft.h:591 | n[n1] | Leader | Active | Replicated on leader n[n1]: 3 (0 hooks)
<RaftDriver>  n[n1]->>n[n1]: [ledger] appending: 2.3=[28 bytes] {"data":"NDI=",
2023-12-19T21:59:35.735318Z        100 [debug] ../src/consensus/aft/impl/state.h:39 | Updating view to: 2 at version: 3
2023-12-19T21:59:35.735322Z        100 [debug] ../src/consensus/aft/impl/state.h:56 | Resulting views: 1, 1
libc++abi: terminating due to uncaught exception of type std::out_of_range: map::at:  key not found
Aborted (core dumped)
```

```tla
Error: Invariant LogMatchingInv is violated.
Error: The behavior up to this point is:
State 1: <MCInit line 103, col 5 to line 110, col 65 of module MCccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 1) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 0)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>> @@ n2 :> <<>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<>>)

State 2: <MCClientRequest(n1) line 65, col 5 to line 66, col 27 of module MCccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 1) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 0)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<>>)

State 3: <AppendEntries(n1,n2) line 551, col 5 to line 581, col 88 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 2) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 0, prevLogTerm |-> 2, entries |-> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 0)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<>>)

State 4: <RcvUpdateTerm(n2,n1) line 997, col 5 to line 999, col 42 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 2) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 0, prevLogTerm |-> 2, entries |-> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> << >>)

State 5: <RcvAppendEntriesRequest(n2,n1) line 1016, col 5 to line 1019, col 58 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 2) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<[dest |-> n1, source |-> n2, term |-> 2, type |-> AppendEntriesResponse, success |-> TRUE, lastLogIndex |-> 1]>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 6: <RcvDropIgnoredMessage(n1,n2) line 990, col 5 to line 992, col 48 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 2) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 7: <AppendEntries(n1,n2) line 551, col 5 to line 581, col 88 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 3) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 1, prevLogTerm |-> 2, entries |-> <<[contentType |-> Signature, term |-> 2]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 8: <RcvAppendEntriesRequest(n2,n1) line 1016, col 5 to line 1019, col 58 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 3) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<[dest |-> n1, source |-> n2, term |-> 2, type |-> AppendEntriesResponse, success |-> TRUE, lastLogIndex |-> 2]>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 9: <RcvDropIgnoredMessage(n1,n2) line 990, col 5 to line 992, col 48 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 3) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 10: <AppendEntries(n1,n2) line 551, col 5 to line 581, col 88 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 4) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 2, prevLogTerm |-> 2, entries |-> <<[contentType |-> Entry, term |-> 2, request |-> 42]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 11: <CheckQuorum(n1) line 723, col 5 to line 725, col 112 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Follower @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 4) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 2, prevLogTerm |-> 2, entries |-> <<[contentType |-> Entry, term |-> 2, request |-> 42]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 12: <MCTimeout(n1) line 49, col 5 to line 60, col 21 of module MCccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Candidate @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> n1 @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 4) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 2, prevLogTerm |-> 2, entries |-> <<[contentType |-> Entry, term |-> 2, request |-> 42]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {n1} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 13: <BecomeLeader(n1) line 585, col 5 to line 599, col 56 of module ccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> n1 @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 3 @@ n2 :> 3) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 2, prevLogTerm |-> 2, entries |-> <<[contentType |-> Entry, term |-> 2, request |-> 42]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {n1} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<{n1}>>)

State 14: <ChangeConfiguration(n1) line 660, col 5 to line 661, col 51 of module ccfraft>
/\ reconfigurationCount = 1
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {n1}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> n1 @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 3 @@ n2 :> 3) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<>> @@ n2 :> <<[commitIndex |-> 2, dest |-> n2, source |-> n1, term |-> 2, type |-> AppendEntriesRequest, prevLogIndex |-> 2, prevLogTerm |-> 2, entries |-> <<[contentType |-> Entry, term |-> 2, request |-> 42]>>]>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Reconfiguration, term |-> 2, configuration |-> {n2}]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2]>>)
/\ votesGranted = (n1 :> {n1} @@ n2 :> {})
/\ configurations = (n1 :> (1 :> {n1} @@ 3 :> {n2}) @@ n2 :> <<{n1}>>)

State 15: <RcvAppendEntriesRequest(n2,n1) line 1016, col 5 to line 1019, col 58 of module ccfraft>
/\ reconfigurationCount = 1
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {n1}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> n1 @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 3 @@ n2 :> 3) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 2)
/\ messages = (n1 :> <<[dest |-> n1, source |-> n2, term |-> 2, type |-> AppendEntriesResponse, success |-> TRUE, lastLogIndex |-> 3]>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 2)
/\ log = (n1 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Reconfiguration, term |-> 2, configuration |-> {n2}]>> @@ n2 :> <<[contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}], [contentType |-> Signature, term |-> 2], [contentType |-> Entry, term |-> 2, request |-> 42]>>)
/\ votesGranted = (n1 :> {n1} @@ n2 :> {})
/\ configurations = (n1 :> (1 :> {n1} @@ 3 :> {n2}) @@ n2 :> <<{n1}>>)
```

```
start_node,n1
emit_signature,2
# MCClientRequest MCccfraft:65
replicate,2,42
assert_commit_idx,n1,2
# AppendEntries ccfraft:551
dispatch_all
assert_commit_idx,n1,2
# RcvUpdateTerm ccfraft:997
# Noop
assert_commit_idx,n1,2
# RcvAppendEntriesRequest ccfraft:1016
# Noop
assert_commit_idx,n1,2
# RcvDropIgnoredMessage ccfraft:990
# Noop
assert_commit_idx,n1,2
# AppendEntries ccfraft:551
dispatch_all
assert_commit_idx,n1,2
# RcvAppendEntriesRequest ccfraft:1016
# Noop
assert_commit_idx,n1,2

### assert below causes code dump because n2 doesn't exist.

assert_commit_idx,n2,2
# RcvDropIgnoredMessage ccfraft:990
# Noop
assert_commit_idx,n1,2
assert_commit_idx,n2,2
# AppendEntries ccfraft:551
dispatch_all
assert_commit_idx,n1,2
assert_commit_idx,n2,2
# CheckQuorum ccfraft:723
periodic_one,n1,110
assert_commit_idx,n1,2
assert_commit_idx,n2,2
# MCTimeout MCccfraft:49
periodic_one,n1,110
assert_commit_idx,n1,2
assert_commit_idx,n2,2
# BecomeLeader ccfraft:585
# Noop
assert_commit_idx,n1,2
assert_commit_idx,n2,2
# ChangeConfiguration ccfraft:660
# Noop
assert_commit_idx,n1,2
assert_commit_idx,n2,2
# RcvAppendEntriesRequest ccfraft:1016
# Noop
assert_commit_idx,n1,2
assert_commit_idx,n2,2
```

achamayou commented:
@lemmy so this first issue is that the mapping ignores `ChangeConfiguration`, because the expectation was that it would translate `ChangeConfigurationInt` instead, given that:

```
    \E newConfiguration \in SUBSET(Servers \ removedFromConfiguration) :
        ChangeConfigurationInt(i, newConfiguration)
```

But `ChangeConfigurationInt` doesn't seem to make it to the trace, I guess because it's a function called by the action and not the action itself. That's a problem, because logging `ChangeConfiguration` alone and its parameter `i` is obviously insufficient. I guess we need to diff the pre and post states to find the new configuration in this case. 

The second issue is that we should use `trust_node`, rather than `replicate_new_configuration`, since that does the creation and trusting, that's my mistake.

The third issue is that this performing complete replacement, not just a node addition, and that we don't have good way to do that in the driver now.

achamayou commented:
Found another mismatch, before even getting to the reconfig:

```
Finished computing initial states: 1 distinct state generated at 2023-12-20 11:36:46.
Error: Invariant LogMatchingInv is violated.
Error: The behavior up to this point is:
State 1: <Initial predicate>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 1) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 0)
/\ log = ( n1 :>
      << [contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}],
         [contentType |-> Signature, term |-> 2] >> @@
  n2 :> <<>> )
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<>>)

State 2: <MCClientRequest(n1) line 65, col 5 to line 66, col 27 of module MCccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 1) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 0)
/\ log = ( n1 :>
      << [contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}],
         [contentType |-> Signature, term |-> 2],
         [contentType |-> Entry, term |-> 2, request |-> 42] >> @@
  n2 :> <<>> )
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<>>)

State 3: <MCSignCommittableMessages(n1) line 73, col 5 to line 74, col 37 of module MCccfraft>
/\ reconfigurationCount = 0
/\ committableIndices = (n1 :> {4} @@ n2 :> {})
/\ removedFromConfiguration = {}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 1) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
/\ messages = (n1 :> <<>> @@ n2 :> <<>>)
/\ currentTerm = (n1 :> 2 @@ n2 :> 0)
/\ log = ( n1 :>
      << [contentType |-> Reconfiguration, term |-> 2, configuration |-> {n1}],
         [contentType |-> Signature, term |-> 2],
         [contentType |-> Entry, term |-> 2, request |-> 42],
         [contentType |-> Signature, term |-> 2] >> @@
  n2 :> <<>> )
/\ votesGranted = (n1 :> {} @@ n2 :> {})
/\ configurations = (n1 :> <<{n1}>> @@ n2 :> <<>>)

State 4: <ChangeConfiguration(n1) line 660, col 5 to line 661, col 51 of module ccfraft>
/\ reconfigurationCount = 1
/\ committableIndices = (n1 :> {4} @@ n2 :> {})
/\ removedFromConfiguration = {n1}
/\ state = (n1 :> Leader @@ n2 :> None)
/\ matchIndex = (n1 :> (n1 :> 0 @@ n2 :> 0) @@ n2 :> (n1 :> 0 @@ n2 :> 0))
/\ votedFor = (n1 :> Nil @@ n2 :> Nil)
/\ nextIndex = (n1 :> (n1 :> 1 @@ n2 :> 1) @@ n2 :> (n1 :> 1 @@ n2 :> 1))
/\ commitIndex = (n1 :> 2 @@ n2 :> 0)
```

The generated assert for the end of state 4 is for commit index 2 on n1:

```
start_node,n1
emit_signature,2
# MCClientRequest MCccfraft:65
replicate,2,42
assert_commit_idx,n1,2
# MCSignCommittableMessages MCccfraft:73
emit_signature,2
assert_commit_idx,n1,2
```

But it is 4 in the implementation, since 4 is committable and the whole config ({n1}) is aware.

achamayou commented:
@lemmy I've fixed reconfiguration by tweaking `replicate_new_configuration` to create nodes when they are missing rather than throw, and added logic that fetches the last config for the node from the post state. That seems work ok on a scenario similar to the one you pasted, but the issue with commit asserts remains.

lemmy commented:
Simulate ccfraft and feed to raft_driver:

```shell
root@codespaces-a47fa6:/workspaces/CCF/tla/consensus# TS=$(date +%s) && java -jar ~/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2023.12.2006/tools/tla2tools.jar -note SIMccfraft.tla -simulate num=1 -dumptrace json SIMccfraft-$TS.json > SIMccfraft-$TS.out ; python3 ../trace2scen.py SIMccfraft-$TS.json > SIMccfraft-$TS.rt ; ../../build/raft_driver SIMccfraft-$TS.rt
```

```diff
diff --git a/tla/consensus/SIMccfraft.cfg b/tla/consensus/SIMccfraft.cfg
index b920aaba8..348c56720 100644
--- a/tla/consensus/SIMccfraft.cfg
+++ b/tla/consensus/SIMccfraft.cfg
@@ -54,6 +54,7 @@ POSTCONDITION
     WriteStatsFile
 
 INVARIANTS
+    Foo
     LogInv
     MoreThanOneLeaderInv
     CandidateTermNotInLogInv
diff --git a/tla/consensus/SIMccfraft.tla b/tla/consensus/SIMccfraft.tla
index bfacc4e7f..6d13789f9 100644
--- a/tla/consensus/SIMccfraft.tla
+++ b/tla/consensus/SIMccfraft.tla
@@ -39,7 +39,7 @@ SIMNext ==
           [] OTHER -> IF ENABLED Forward THEN Forward ELSE Next
 
 SIMSpec ==
-    Init /\ [][SIMNext]_vars
+    Init /\ [][Next]_vars \* Next and not SIMNext to get proper action names & params.
 
 \* The state constraint  StopAfter  stops TLC after the alloted
 \* time budget is up, unless TLC encounteres an error first.
@@ -48,6 +48,8 @@ StopAfter ==
     (* The smoke test has a time budget of 20 minutes. *)
     IN TLCSet("exit", TLCGet("duration") > timeout)
 
+Foo ==
+    TLCGet("level") < 100
 =============================================================================
 
 ------------------------------- MODULE SIMPostCondition -------------------------------
diff --git a/tla/trace2scen.py b/tla/trace2scen.py
index d4e9ddb66..2811a5c09 100644
--- a/tla/trace2scen.py
+++ b/tla/trace2scen.py
@@ -32,6 +32,12 @@ MAP = {
     "RejectAppendEntriesRequest": lambda _, __, ___: ["# Noop"],
     "ReturnToFollowerState": lambda _, __, ___: ["# Noop"],
     "AppendEntriesAlreadyDone": lambda _, __, ___: ["# Noop"],
+    "RcvUpdateTerm": lambda _, __, ___: ["# Noop"],
+    "RcvAppendEntriesRequest": lambda _, __, ___: ["# Noop"],
+    "RcvAppendEntriesResponse": lambda _, __, ___: ["# Noop"],
+    "RcvDropIgnoredMessage": lambda _, __, ___: ["# Noop"],
+    "RcvRequestVoteRequest": lambda _, __, ___: ["# Noop"],
+    "RcvRequestVoteResponse": lambda _, __, ___: ["# Noop"],
 }
 
 def post_commit(post):
```

lemmy commented:
> The first issue is that the mapping ignores `ChangeConfiguration`, because the (unreasonably optimistic) expectation was that it would translate `ChangeConfigurationInt` instead, given that:
> 
> ```
>     \E newConfiguration \in SUBSET(Servers \ removedFromConfiguration) :
>         ChangeConfigurationInt(i, newConfiguration)
> ```
> 
> But `ChangeConfigurationInt` doesn't seem to make it to the trace, I guess because it's a function called by the action and not the action itself. That's annoying, because logging `ChangeConfiguration` alone and its parameter `i` is obviously insufficient. I guess we need to diff the pre and post states to find the new configuration in this case.

This (syntactic) change causes `ChangeConfigurationInt` with parameters `i` and `newConfiguration` to be included in the trace.

```diff
diff --git a/tla/consensus/ccfraft.tla b/tla/consensus/ccfraft.tla
index 7d5fae84c..af1eefd69 100644
--- a/tla/consensus/ccfraft.tla
+++ b/tla/consensus/ccfraft.tla
@@ -635,6 +635,8 @@ SignCommittableMessages(i) ==
 ChangeConfigurationInt(i, newConfiguration) ==
     \* Only leader can propose changes
     /\ state[i] = Leader
+    \* Configuration is not a previously removed configuration.
+    /\ newConfiguration \notin removedFromConfiguration
     \* Configuration is non empty
     /\ newConfiguration /= {}
     \* Configuration is a proper subset of the Servers
@@ -657,7 +659,7 @@ ChangeConfigurationInt(i, newConfiguration) ==
                     leaderVars, commitIndex, committableIndices>>
 
 ChangeConfiguration(i) ==
-    \E newConfiguration \in SUBSET(Servers \ removedFromConfiguration) :
+    \E newConfiguration \in SUBSET(Servers) :
         ChangeConfigurationInt(i, newConfiguration)
 
 \* Leader i advances its commitIndex to the next possible Index.
```

lemmy commented:
### What are the use cases for *lockstep debugging* (ccfraft.tla > raft.h > Traececcfraft.tla) of the spec and the impl (via `raft_driver`)?

+ Lockstep debugging (or co-debugging) is an interesting technical challenge. :-)
+ A single breakpoint in the spec or impl will also halt the "other side".
+ Spec has depth/level aware breakpoints, i.e., advance behavior up to diameter N.
+ Inspect the full state of the impl at state `s` beyond what is serialized in the json output. 
   => Workaround:  One can simply run the generated raft scenario up to the point where the impl is in the desired state.
+ The `Traceccfraft.tla` part `raft.h > Traceccfraft.tla` provides the global system perspective, whereas the raft.h debugger provides a node-local perspective. (Question: How do we debug multiple nodes with `raft_driver`?)

It appears as if a main benefit of co-debugging/lockstep debugging is locked away by the fact that the driver cannot reverse the impl.  Otherwise, one could interactively explore the state space at every state of a behavior.

#### Technical considerations

1) Emit the next raft scenario command 
   a) When generating a complete scenario from a TLA+ trace
   b) One by one when creating a TLA+ trace with e.g. the TLA+ debugger
2) Validate the resulting json output
   a) With `Traceccfraft.tla` after the json log has been serialized
   b) Interactively, when moving to the next state with e.g. the TLA+ debugger

"Batch" is trivial: Write trace and print raft scenario command with an Alias or a (Python) script.  Verify the json output with existing `Traceccfraft.tla`.  Interactive is more challenging: Emit 1b) with `TLCDefer` as part of the next-state relation, and verify 2b in an action-constraint.

##### TLA+/TLC infra

- `dumptrace`
- TLCDefer
- Alias
- State- & Action-Constraints
- Hyperproperties, i.e., observational determinism (see e.g. https://www.hillelwayne.com/post/hyperproperties/ and https://www.youtube.com/watch?v=K4PcSNOs9ok).  Let `DS` be the debug (ccfraft) spec and `TS` the `Traceccfraft` spec.  A HyperSpec defines `Next == DS!Next /\ TS!Next`, where `DS!Next` emits the next raft scenario command, and `TS!Next` reads the json output of the raft driver.

achamayou commented:
@lemmy thank you for making the change in #5880, I've update the conversion, and also added the missing actions you spotted when running the sim. It's a little annoying that the default SIM traces don't work, and that changing the Next is necessary, but the improved coverage is desirable.

> (Question: How do we debug multiple nodes with raft_driver?)

raft_driver is single threaded, nodes are separate objects but all execution is serialised.

> It appears as if a main benefit of co-debugging/lockstep debugging is locked away by the fact that the driver cannot reverse the impl.

Agreed, though as discussed before the new year, making the driver able to reverse the state of each node is probably achievable: nearly all of it already supports trivial json ser/deser. We would also need to put the messages back on the queues, and restore time elapsed. I don't know that we want to go much further before we complete #5822 at least.

For now, the intermediate result this gives us is: "A validation trace can (mostly) be fed back to the driver for investigation".

eddyashton commented on tla/trace2scen.py:
```diff
@@ -0,0 +1,73 @@
+    "SignCommittableMessages": lambda ctx, pre, post: ["emit_signature", term(ctx, pre)],
+    "MCSignCommittableMessages": lambda ctx, pre, post: ["emit_signature", term(ctx, pre)],
+    "ChangeConfigurationInt": lambda ctx, pre, post: ["replicate_new_configuration", term(ctx, pre), *ctx["newConfiguration"]],
+    "AdvanceCommitIndex": noop,
```
Can we make these print `# Noop (AdvanceCommitIndex)`, rather than identical `# Noop`s? Would mean we get a clearer correspondence to each trace action, without counting noops.

If they dominate real generated scenarios, I'd also be tempted to just omit lines for these entirely.

achamayou replied:
We can, but they each immediately follow a comment that contains the action, the line etc already (see comment()), so it's a bit redundant. The rationale for comment + actions is that there can be more than one action emitted (like for bootstrap).

There are quite a few, but I think it's good to keep them for clarity.

eddyashton commented on tla/trace2scen.py:
```diff
@@ -0,0 +1,73 @@
+    return ["# Noop"]
+
+MAP = {
+    "ClientRequest": lambda ctx, pre, post: ["replicate", term(ctx, pre), "42"],
```
Is there a reason these return lists, joined on line 58, rather than formatting directly?

eg

```suggestion
    "ClientRequest": lambda ctx, pre, post: f"replicate,{term(ctx,pre)},42",
```

If a trace action ever results in multiple scenario steps, then we'll want a list of results then to be joined later, and then the nested lists will be more confusing.

achamayou replied:
I was mostly trying to keep everything structured until a final flattening step, to facilitate any additional passes we may need to do. But we don't have any, and we could just as easily format directly here.

eddyashton commented:
Can we get traces from the existing (matching) scenarios, and see what scenarios this generates from them?

lemmy commented:
> Can we get traces from the existing (matching) scenarios, and see what scenarios this generates from them?

We have only two scenarios compatible with the new bootstrapping logic.  @achamayou does it make sense to port all deprecated scenarios?

achamayou commented:
> > Can we get traces from the existing (matching) scenarios, and see what scenarios this generates from them?
> 
> We have only two scenarios compatible with the new bootstrapping logic. @achamayou does it make sense to port all deprecated scenarios?

I've been trying to do that (#5839), but have not found enough time so far.

The traces won't match, because all information related to message dispatch and periodic/time elapsed is lost right now. I don't think we can get them to match without substantially reworking either the spec or the driver, as discussed.
